### PR TITLE
Command line tool for generating test handler code from templates

### PR DESCRIPTION
**Pull request by shimritproj**

### CODE REVIEW

File: cmd/cli/main.go [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+type handler struct {
+	X string
+}
+
+var (
+	pathrelativetoroot string
+	handlerDirectory   string
+	handlername        string
+	rootCmd            = &cobra.Command{
+		Use:   "kind",
```
Comment (jun-rh) [inline cmd/cli/main.go:22]: Update this

File: cmd/cli/main.go [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+		Short: "kind is a tool for managing local Kubernetes clusters",
+		Long:  "kind creates and manages local Kubernetes clusters using Docker container 'nodes'",
+		Run: func(cmd *cobra.Command, args []string) {
+
+			handlername = args[len(args)-1]
```
Comment (jun-rh) [inline cmd/cli/main.go:27]: Use subcommands to do this.

**Review by jun-rh — CHANGES_REQUESTED** [review]
Fix lint error
Update readme

File: cmd/cli/main.go [hunk 3]
```diff
@@ -0,0 +1,82 @@
+package main
```
Comment (jun-rh) [inline cmd/cli/main.go:1]: are we really gonna call the binary "cli" or "tnf"?

File: cmd/tnf_cobra/main.go [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+	X string
+}
+
+var (
+	pathrelativetoroot string
+	handlerDirectory   string
+	handlername        string
+
+	rootCmd = &cobra.Command{
+		Use:   "rootCmd",
```
Comment (jun-rh) [inline cmd/tnf_cobra/main.go:23]: tnf

File: cmd/tnf_cobra/main.go [hunk 5]
```diff
@@ -0,0 +1,97 @@
+package main
```
Comment (jun-rh) [inline cmd/tnf_cobra/main.go:1]: folder name should be just tnf

**Review by jun-rh — CHANGES_REQUESTED** [review]
Also, fix the lint error and update DEVELOPING.md

File: cmd/tnf_cobra/main.go [hunk 6]
```diff
@@ -0,0 +1,97 @@
...
+	}
+
+	generate = &cobra.Command{
+		Use:   "generate",
+		Short: "generator tool for various tnf artifacts, such as handler code, catalog etc",
+	}
+
+	handler = &cobra.Command{
+		Use:   "handler",
+		Short: "adding new handler",
```
Comment (jun-rh) [inline cmd/tnf_cobra/main.go:34]: Try improve this a bit

File: cmd/tnf_cobra/main.go [hunk 7]
```diff
@@ -0,0 +1,97 @@
...
+		Run:   addingHandler,
+	}
+)
+
+func addingHandler(cmd *cobra.Command, args []string) {
```
Comment (jun-rh) [inline cmd/tnf_cobra/main.go:39]: Function name is a bit confusing. How about "generateHandlerFiles"?

File: DEVELOPING.md [hunk 8]
```diff
@@ -604,6 +604,11 @@ The same result could be achieved using a Go Template:
 oc get pod %s -n %s -o go-template='{{len .spec.containers}}{{"\n"}}'
 ```
 
+## Adding new handler
+
+When we want to creat new handler, we need to write this command: "go run main.go generate handler <<"new handler">>", we need to running this command when we in folder "cmd/tnf/"
```
Comment (jun-rh) [inline DEVELOPING.md:609]: "tnf" instead of "go run main.go"

File: DEVELOPING.md [hunk 9]
```diff
@@ -604,6 +604,11 @@ The same result could be achieved using a Go Template:
...
+The result is folder "new handler" that include 3 files by handler template.
```
Comment (jun-rh) [inline DEVELOPING.md:610]: list the 3 files/path and their purpose. Think of this as for a new developer joining the project to read.

File: .gitignore [hunk 10]
```diff
@@ -20,3 +20,5 @@ pkg/tnf/reel/mocks/mock_reel.go
 # NOTE: mock_expect.go is tracked, as it is generated from source externally.
 pkg/tnf/interactive/mocks/mock_spawner.go
 temp/
+go.mod
```
Comment (jun-rh) [inline .gitignore:23]: Is this intentional?

File: cmd/tnf/main.go [hunk 11]
```diff
@@ -0,0 +1,95 @@
...
+	Handlername string
+}
+
+var (
+	pathrelativetoroot string
+	handlerDirectory   string
+	handlername        string
+
+	rootCmd = &cobra.Command{
+		Use:   "rootCmd",
```
Comment (jun-rh) [inline cmd/tnf/main.go:23]: tnf, not rootCmd

File: cmd/tnf/main.go [hunk 12]
```diff
@@ -0,0 +1,95 @@
...
+func createfile(pathfile, namefile string, myhandler myHandler, newHandlerDirectory string) {
+	ftpl, err := template.ParseFiles(pathfile)
+	if err != nil {
+		log.Fatalln(err)
+	}
+
+	temp := path.Join(newHandlerDirectory, namefile)
+	f, err := os.Create(temp)
+	if err != nil {
+		panic(err)
```
Comment (jun-rh) [inline cmd/tnf/main.go:75]: why panic? log the error and return

File: pkg/tnf/handlers/handler_template/template.tmpl [hunk 13]
```diff
@@ -0,0 +1,104 @@
...
+type {{ .Handlername }} struct {
+	result  int
+	timeout time.Duration
+	args    []string
+	// The hostname
+	hostname string
+}
+
+const (
+	// Command is the command name for the unix "hostname" command.
```
Comment (HamadiSE) [inline pkg/tnf/handlers/handler_template/template.tmpl:38]: this is not repeated accross all the files.
If something is not common between all the tests, then it should be replaced with //TO_DO or a generic comment

File: cmd/tnf/main.go [hunk 14]
```diff
@@ -0,0 +1,95 @@
...
+	pathfile = path.Join(handlerDirectory, "handler_template", "template_test.tmpl")
+	namefile = "" + handlername + "_test.tmpl"
+	createfile(pathfile, namefile, myhandler, newHandlerDirectory) // here creating file by template_test.tmpl
+
+	pathfile = path.Join(handlerDirectory, "handler_template", "template.tmpl")
+	namefile = "" + handlername
+	createfile(pathfile, namefile, myhandler, newHandlerDirectory) // here creating file by template.tmpl
+}
+
+func createfile(pathfile, namefile string, myhandler myHandler, newHandlerDirectory string) {
```
Comment (jun-rh) [inline cmd/tnf/main.go:66]: this function should return error so that the caller can decide whether to proceed.

File: pkg/tnf/handlers/handler_template/template.tmpl [hunk 15]
```diff
@@ -0,0 +1,104 @@
...
+func (h *{{ .Handlername }}) ReelEOF() {
+}
+
+// GetHostname returns the extracted hostname, if one is extracted.
+func (h *{{ .Handlername }}) GetHostname() string {
+	return h.hostname
+}
+
+// NewHostname creates a new `Hostname` test which runs the "hostname" command.
+func NewHostname(timeout time.Duration) *{{ .Handlername }} {
```
Comment (HamadiSE) [inline pkg/tnf/handlers/handler_template/template.tmpl:95]: same here. this is not something that should be in the template.
The template should included only the generic/programmable part that's shareable by all the test.

File: pkg/tnf/handlers/handler_template/template.tmpl [hunk 16]
```diff
@@ -0,0 +1,104 @@
+// Copyright (C) 2020-2021 Red Hat, Inc.
```
Comment (HamadiSE) [inline pkg/tnf/handlers/handler_template/template.tmpl:1]: the name of the file should be updated. Is this the test template ? or the handler template ?

File: pkg/tnf/handlers/handler_template/template_test.tmpl [hunk 17]
```diff
@@ -0,0 +1,77 @@
...
+	"github.com/test-network-function/test-network-function/pkg/tnf/handlers/hostname"
+	"github.com/test-network-function/test-network-function/pkg/tnf/identifier"
+)
+
+const (
+	testTimeoutDuration = time.Second * 2
+)
+
+func {{ .Handlername }}_Args(t *testing.T) {
+	h := hostname.NewHostname(testTimeoutDuration)
```
Comment (HamadiSE) [inline pkg/tnf/handlers/handler_template/template_test.tmpl:34]: same thing here. the generated test has nothing to do with hostname test

File: pkg/tnf/handlers/handler_template/template_test.tmpl [hunk 18]
```diff
@@ -0,0 +1,77 @@
...
+	assert.Equal(t, []string{"hostname"}, h.Args())
+}
+
+func TestHostname_GetIdentifier(t *testing.T) {
+	h := hostname.NewHostname(testTimeoutDuration)
```
Comment (HamadiSE) [inline pkg/tnf/handlers/handler_template/template_test.tmpl:39]: ditto..

File: pkg/tnf/handlers/handler_template/template_test.tmpl [hunk 19]
```diff
@@ -0,0 +1,77 @@
...
+	assert.Equal(t, identifier.HostnameIdentifier, h.GetIdentifier())
+}
+
+func TestHostname_ReelFirst(t *testing.T) {
+	h := hostname.NewHostname(testTimeoutDuration)
```
Comment (HamadiSE) [inline pkg/tnf/handlers/handler_template/template_test.tmpl:44]: ditto

File: pkg/tnf/handlers/handler_template/template_test.tmpl [hunk 20]
```diff
@@ -0,0 +1,77 @@
...
+	step := h.ReelFirst()
+	assert.Equal(t, "", step.Execute)
+	assert.Equal(t, []string{hostname.SuccessfulOutputRegex}, step.Expect)
+	assert.Equal(t, testTimeoutDuration, step.Timeout)
+}
+
+func TestHostname_ReelEof(t *testing.T) {
+	h := hostname.NewHostname(testTimeoutDuration)
+	// just ensures lack of panic
```
Comment (HamadiSE) [inline pkg/tnf/handlers/handler_template/template_test.tmpl:53]: ditto

File: pkg/tnf/handlers/handler_template/template_test.tmpl [hunk 21]
```diff
@@ -0,0 +1,77 @@
...
+	h := hostname.NewHostname(testTimeoutDuration)
+	matchHostname := "testHostname"
+	step := h.ReelMatch("", "", matchHostname)
+	assert.Nil(t, step)
+	assert.Equal(t, matchHostname, h.GetHostname())
+	assert.Equal(t, tnf.SUCCESS, h.Result())
+}
+
+func TestNewHostname(t *testing.T) {
+	h := hostname.NewHostname(testTimeoutDuration)
```
Comment (HamadiSE) [inline pkg/tnf/handlers/handler_template/template_test.tmpl:74]: ditto

File: pkg/tnf/handlers/handler_template/handler.tmpl [hunk 22]
```diff
@@ -0,0 +1,96 @@
...
+// ReelMatch parses the hostname output and set the test result on match.
+// Returns no step; the test is complete.
+func (h *{{ .Handlername }}) ReelMatch(_, _, match string) *reel.Step {
+	h.hostname = match
+	h.result = tnf.SUCCESS
+	return nil
+}
+
+// ReelTimeout does nothing;  hostname requires no explicit intervention for a timeout.
+func (h *{{ .Handlername }}) ReelTimeout() *reel.Step {
```
Comment (HamadiSE) [inline pkg/tnf/handlers/handler_template/handler.tmpl:76]: there's few places where hostname is still mentionned.
The code still needs more cleaning to remove out non-generic parts and replace them with Action/TODO..

**Review by HamadiSE — CHANGES_REQUESTED** [review]

File: .gitignore [hunk 23]
```diff
@@ -20,3 +20,4 @@ pkg/tnf/reel/mocks/mock_reel.go
...
```
Comment (jun-rh) [inline .gitignore:23]: Why is this added?

File: cmd/tnf/main.go (hunk 11 above)
Comment (jun-rh) [reply cmd/tnf/main.go]: still rootCmd?

File: pkg/tnf/handlers/handler_template/handler.tmpl [hunk 24]
```diff
@@ -0,0 +1,96 @@
...
+type {{ .Handlername }} struct {
+	result  int
+	timeout time.Duration
+	args    []string
+	// The hostname
+	hostname string
+}
+
+const (
+	// Command is the command name for the unix "hostname" command.
```
Comment (jun-rh) [inline pkg/tnf/handlers/handler_template/handler.tmpl:33]: hostname

File: cmd/tnf/main.go [hunk 25]
```diff
@@ -0,0 +1,109 @@
...
+	"fmt"
+	"log"
+	"os"
+	"path"
+	"text/template"
+
+	"github.com/spf13/cobra"
+)
+
+const success = "success"
```
Comment (jun-rh) [inline cmd/tnf/main.go:14]: No need for this. Returning nil means success. This is common practice in go.

File: pkg/tnf/handlers/handler_template/handler.tmpl [hunk 26]
```diff
@@ -18,15 +18,10 @@ package {{ .Handlername }}
 
 import (
 )
-//TODO
-// Hostname provides a hostname test implemented using command line tool "hostname".
-/*
+
+//TODO:
 type {{ .Handlername }} struct {
```
Comment (HamadiSE) [inline pkg/tnf/handlers/handler_template/handler.tmpl:23]: in here
``
timeout time.Duration
	args    []string
``
Those parameters are always present. You can keep them in the template

File: pkg/tnf/handlers/handler_template/handler.tmpl [hunk 27]
```diff
@@ -36,61 +31,45 @@ const (
...
 func (h *{{ .Handlername }}) Args() []string {
-	return h.args
+	//fill the stub
 }
 
 // GetIdentifier returns the tnf.Test specific identifier.
 func (h *{{ .Handlername }}) GetIdentifier() identifier.Identifier {
-	return identifier.HostnameIdentifier
+	//fill the stub
 }
```
Comment (HamadiSE) [inline pkg/tnf/handlers/handler_template/handler.tmpl:39]: can you update the comment to something like :
`// TODO create identifier.{{ .Handername}}Identifier; uncomment line below              

// return identifier.{{ .Handername}}Identifier
`

File: pkg/tnf/handlers/handler_template/handler.tmpl [hunk 28]
```diff
@@ -36,61 +31,45 @@ const (
...
 
 // Timeout return the timeout for the test.
 func (h *{{ .Handlername }}) Timeout() time.Duration {
-	return h.timeout
+	//fill the stub
```
Comment (HamadiSE) [inline pkg/tnf/handlers/handler_template/handler.tmpl:43]: same thing here. The ``timeout`` is common across all the files. You can keep this function in the template

File: pkg/tnf/handlers/handler_template/handler.tmpl [hunk 29]
```diff
@@ -36,61 +31,45 @@ const (
...
 }
 
-// ReelFirst returns a step which expects an hostname summary for the given device.
 func (h *{{ .Handlername }}) ReelFirst() *reel.Step {
-	return &reel.Step{
-		Expect:  []string{SuccessfulOutputRegex},
-		Timeout: h.timeout,
-	}
+	//fill the stub
 }
```
Comment (HamadiSE) [inline pkg/tnf/handlers/handler_template/handler.tmpl:53]: in here. the only the Expect is not common.
I think it should be something like
```
// ReelFirst
func (h *{{ .Handlername }}) ReelFirst() *reel.Step {
	return &reel.Step{
		Expect: []string{""}, //TODO : pass the list of possible regex in here
		Timeout: h.timeout,
	}
}
``

File: pkg/tnf/handlers/handler_template/handler.tmpl [hunk 30]
```diff
@@ -36,61 +31,45 @@ const (
...
-	h.result = tnf.SUCCESS
-	return nil
+	//fill the stub
 }
 
-// ReelTimeout does nothing;  hostname requires no explicit intervention for a timeout.
 func (h *{{ .Handlername }}) ReelTimeout() *reel.Step {
-	return nil
+	//fill the stub
 }
```
Comment (HamadiSE) [inline pkg/tnf/handlers/handler_template/handler.tmpl:61]: you should keep return nil with another comment asking the user to add any code if necessary

**Review by HamadiSE — CHANGES_REQUESTED** [review]

File: pkg/tnf/handlers/handler_template/handler.tmpl [hunk 31]
```diff
@@ -36,61 +31,45 @@ const (
...
 
-// ReelEOF does nothing;  hostname requires no explicit intervention for EOF.
 func (h *{{ .Handlername }}) ReelEOF() {
+	//fill the stub
 }
 
-// GetHostname returns the extracted hostname, if one is extracted.
 func (h *{{ .Handlername }}) GetHostname() string {
```
[the remark this answers is not in the document]
Comment (shimritproj) [reply pkg/tnf/handlers/handler_template/handler.tmpl]: why? I wrote this function: 
func (h *{{ .Handlername }}) Get{{ .Handlername }}() string {
	//fill the stub
}
this OK?

File: pkg/tnf/handlers/handler_template/handler.tmpl (hunk 26 above)
Comment (shimritproj) [reply pkg/tnf/handlers/handler_template/handler.tmpl:23]: this kept in template.

File: cmd/tnf/main.go [hunk 32]
```diff
@@ -2,7 +2,6 @@ package main
 
 import (
```
[the remark this answers is not in the document]
Comment (jun-rh) [reply cmd/tnf/main.go:3]: @HamadiSE you mean extra functions in handler_test for testing JSON based handler?

[the remark this answers is not in the document]
Comment (HamadiSE) [reply cmd/tnf/main.go:3]: Right. That was Shimrit opinion as well.

**Comment by HamadiSE** [comment]
@shimritproj have you tried generating a a handler, then making sure the code compiles and all UT pass ?

File: pkg/tnf/handlers/handler_template/template.tmpl (hunk 15 above)
Comment (shimritproj) [reply pkg/tnf/handlers/handler_template/template.tmpl]: this is old, I updated this file.

**Comment by HamadiSE** [comment]
few general comments, the generated code does not compile
``
➜  test-network-function git:(shimrit_change) ✗ make
make test
go build -race ./...
pkg/tnf/handlers/Xyz/doc.go:16:1: expected 'package', found '<'
make[1]: *** [test] Error 1
make: *** [build] Error 2
``
Second, it does not generate handlers using the command in Readme file.

File: cmd/tnf/main.go [hunk 33]
```diff
@@ -0,0 +1,107 @@
...
+	handler = &cobra.Command{
+		Use:   "handler",
+		Short: "adding new handler.",
+		RunE:  generateHandlerFiles,
+	}
+)
+
+func generateHandlerFiles(cmd *cobra.Command, args []string) error {
+	handlername = args[0]
+	pathrelativetoroot = path.Join("..", "..")
```
Comment (HamadiSE) [inline cmd/tnf/main.go:41]: this should be : 
**pathrelativetoroot = path.Join("", "")**
are you sure this code work ? I could not make it work on my machine

File: pkg/tnf/handlers/handler_template/handler_test.tmpl [hunk 34]
```diff
@@ -0,0 +1,54 @@
...
+import (
+	"time"
+)
+
+const (
+	testTimeoutDuration = time.Second * 2
+)
+
+//TODO:
+func {{ .Handlername }}_Args{
```
Comment (HamadiSE) [inline pkg/tnf/handlers/handler_template/handler_test.tmpl:28]: // {{ .Handlername }}_Args() 
 func Test{{ .Handlername }}_Args(){
// Todo : write test for function {{ .Handlername }}_Args()
}

**Review by HamadiSE — APPROVED** [review]

**Review by jun-rh — APPROVED** [review]
